Update data-oriented.md for Swan Lake GA (#3323)
ballerina-platform/ballerina-dev-website: learn/why-ballerina/data-oriented.md

```ballerina
import ballerina/io;

type Author record {|
    readonly int id;
    string name;
|};

type Category record {|
    readonly int id;
    string name;
|};

type Book record {|
    readonly int id;
    string title;
    int year;
    float price;
    Author[] authors;
    Category[] categories;
|};

type Sale record {|
    int bookId;
    int qty;
|};

type AuthorTable table<Author> key(id);

type CategoryTable table<Category> key(id);

type BookTable table<Book> key(id);

error onConflictError = error("Key Conflict", message = "record with same key exists.");
```

In the above code, the necessary custom types and variables are created to perform the data manipulation.

```ballerina
function loadAuthors() returns AuthorTable|error {
    json[] authors = [
        {"id": 1, "name": "Giada De Laurentiis"},
        {"id": 2, "name": "J. K. Rowling"},
        {"id": 3, "name": "Henrique C. M. Andrade"},
        {"id": 4, "name": "Buğra Gedik"},
        {"id": 5, "name": "Deepak S. Turaga"}
    ];

>     // Iterates through `authors` JSON array, constructs `Author` records, and collects them into a table.

Reviewer: ```suggestion
    // Iterates through the `authors` JSON array, constructs the `Author` records, and collects them into a table.
```

                           <name>education</name>
                       </category>
                   </categories>`;

>     // Iterates through `categories` xml array, constructs `Category` records, and collects them into a table.

Reviewer: ```suggestion
    // Iterates through the `categories` XML array, constructs the `Category` records, and collects them into a table.
```

    return table key(id) from var category in categories/<category>
        select {
            id: check 'int:fromString((category/**/<id>/*).toString()),
            name: (category/**/<name>/*).toString()
        }
        on conflict onConflictError;
}
```

The above method iterates through a `categories` XML array and constructs a table with the category ID and name. Query expression clauses such as `from`, `select`, and `on conflict` are used in it.

```ballerina
function getCategories(CategoryTable categories, int[] ids) returns Category[] {

>     // Query `categories` for each id, and collects matching categories into an array.

Reviewer: ```suggestion
    // Queries the `categories` of each ID and collects the matching categories into an array.
```

    return from int id in ids
        join Category c in categories on id equals c.id
        select c;
}
```

The above method returns the book category names according to the given category IDs. As written in the query expression, a `join` clause iterates through the `categories` array to find the respective category name according to the ID. 

```ballerina
function getAuthors(AuthorTable authors, int[] ids) returns Author[] {

>     // Query `authors` for each id, and collects matching authors into an array.

Reviewer: ```suggestion
    // Queries the `authors` of each ID and collects the matching authors into an array.
```

    return from int id in ids
        join Author a in authors on id equals a.id
        select a;
}
```

Here, the author names are identified according to the author IDs. As similar to the previous method, query joins are utilized to create the `authors` array. 

```ballerina
function loadBooks() returns BookTable|error {
    map<anydata>[] books = [
        {"id": 1, "title": "Everyday Italian", "year": 2005, "price": 30.00, "authors": [1], "categories": [1]},
        {"id": 2, "title": "Harry Potter", "year": 1997, "price": 29.99, "authors": [2], "categories": [2, 3]},
        {"id": 3, "title": "Fundamentals of Stream Processing", "year": 2014, "price": 123.99, "authors": [3, 4, 5], "categories": [4]},
        {"id": 4, "title": "Fantastic Beasts", "year": 2001, "price": 29.99, "authors": [2], "categories": [2, 3]}
    ];

>     // Iterates through `books` map<anydata> array, constructs `Book` records, and collects them into a table.

Reviewer: ```suggestion
    // Iterates through the `books` map<anydata> array, constructs the `Book` records, and collects them into a table.
```

    return table key(id) from var book in books
        let AuthorTable authors = check loadAuthors()
        let CategoryTable categories = check loadCategories()
        select {
            id: <int>book["id"],
            title: <string>book["title"],
            year: <int>book["year"],
            price: <float>book["price"],
            authors: getAuthors(authors, check book["authors"].cloneWithType()),
            categories: getCategories(categories, check book["categories"].cloneWithType())
        }
        on conflict onConflictError;
}

```

> Now, the author details and book category details exist as individual table values. The book details exist as an array. Further, this array has the necessary keys to identify the respective authors and book categories. Hence, there is a requirement to generate a table that contains the relevant book details, author names, and category names. As shown in the above example, Ballerina query expressions can be utilized to create such table values.  

Reviewer: ```suggestion
Now, the author details and book category details exist as individual table values. The book details exist as an array. Further, this array has the necessary keys to identify the respective authors and book categories. Hence, there is a requirement to generate a table, which contains the relevant book details, author names, and category names. As shown in the above example, Ballerina query expressions can be utilized to create such table values.  
```

```ballerina
function getPopularBooks(stream<Sale> sales, int minSales, int 'limit) returns Book[]|error {

>     // Join each sale value from `sales` stream with retrieved `BookTable`

Reviewer: ```suggestion
    // Join each sale value from the `sales` stream with the retrieved `BookTable`,
```

    // filter books with at least `minSales` number of sales,
    // order filtered books by sales quantity,

>     // and collect `'limit` number of books into an array.

Reviewer: ```suggestion
    // and collect the `'limit` number of books into an array.
```

    return from Sale s in sales
        join Book b in check loadBooks() on s.bookId equals b.id
        where s.qty >= minSales
        order by s.qty descending
        limit 'limit
        select b;
}
```

Now, you have the book details as a table value and there is a requirement to identify the popular books based on the number of sales copies. As shown in the above example, you have to combine the `sales` array and the `books` table to identify the popular books. In this example, Ballerina query clauses such as `join`, `where`, `order by`, and `limit` are utilized to cater to the requirement. 

```ballerina
public function main() returns error? {
    Sale[] sales = [
        {bookId: 1, qty: 150},
        {bookId: 2, qty: 3500},
        {bookId: 3, qty: 1250},
        {bookId: 4, qty: 2100}
    ];

>     // Retrieve 2 most popular books with at least 1000 sales.

Reviewer: ```suggestion
    // Retrieve the two most popular books with at least 1000 sales.
```
